algebra-core/src/curves/cuda/scalar_mul/cpu_gpu_macros.rs

```diff
@@ -0,0 +1,246 @@
+        }
+
+        #[allow(unused_variables)]
+        fn cpu_gpu_load_balance_run_kernel(
+            ctx: &Context,
+            bases_h: &[<Self as ProjectiveCurve>::Affine],
+            exps_h: &[<<Self as ProjectiveCurve>::ScalarField as PrimeField>::BigInt],
+            cuda_group_size: usize,
+            // size of a single job in the queue e.g. 2 << 14
+            job_size: usize,
+            // size of the batch for cpu scalar mul
+            cpu_chunk_size: usize,
+        ) -> Vec<<Self as ProjectiveCurve>::Affine> {
+            #[cfg(feature = "cuda")]
+            {
+                let mut bases_res = bases_h.to_vec();
+                let queue = Mutex::new(bases_res.chunks_mut(job_size).zip(exps_h.chunks(job_size)).peekmore());
+
+                rayon::scope(|s| {
+                    // We launch two concurrent GPU threads that block on waiting for GPU to hide latency
+                    for i in 0..2 {
+                        s.spawn(closure!(move i, ref queue, |_| {
+                            std::thread::sleep(std::time::Duration::from_millis(i * 500));
+                            let mut iter = queue.lock().unwrap();
+                            while let Some((bases, exps)) = iter.next() {
+                                iter.peek();
+                                if iter.peek().is_none() { break; }
+                                let mut proj_res = Self::par_run_kernel_sync(ctx, bases, exps, cuda_group_size, iter);
+                                Self::batch_normalization(&mut proj_res[..]);
+                                bases.clone_from_slice(&proj_res.par_iter().map(|p| p.into_affine()).collect::<Vec<_>>()[..]);
+                                iter = queue.lock().unwrap();
+                            }
+                        }));
+                    }
+
+                    s.spawn(|_| {
+                        std::thread::sleep(std::time::Duration::from_millis(20));
+                        let mut iter = queue.lock().unwrap();
+                        println!("acquired cpu");
+                        while let Some((bases, exps)) = iter.next() {
+                            let exps_mut = &mut exps.to_vec()[..];
+                            rayon::scope(|t| {
+                                for (b, s) in bases.chunks_mut(cpu_chunk_size).zip(exps_mut.chunks_mut(cpu_chunk_size)) {
+                                    t.spawn(move |_| b[..].batch_scalar_mul_in_place(&mut s[..], 4));
+                                }
+                            });
+                            // Sleep to allow other threads to unlock
+                            drop(iter);
+                            println!("unlocked cpu");
+                            std::thread::sleep(std::time::Duration::from_millis(20));
+                            iter = queue.lock().unwrap();
+                            println!("acquired cpu");
+                        }
+                        println!("CPU FINISH");
```

Comment:
This shouldn't appear in a normal run. Either change this to a debug print with logging or remove it.